File: 프로그래머스/lv3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<vector<int>> triangle) 
{
    int answer = 0;
    
    vector<vector<int>> vec;
    vec.push_back(triangle[0]);
    
    for(int i=1; i<triangle.size(); i++)
    {
        vector<int> tmp(i+1);
        
        tmp[0] = vec[i-1][0] + triangle[i][0];
        tmp[0+1] = vec[i-1][0] + triangle[i][0+1];
        for(int t=0; t<vec[i-1].size(); t++)
        {
            tmp[t] = max(tmp[t],vec[i-1][t] + triangle[i][t]);
            
            tmp[t+1] = vec[i-1][t] + triangle[i][t+1];
        }
        vec.push_back(tmp);
    }
    
    answer = vec[vec.size()-1][0];
    for(int i=1; i<vec[vec.size()-1].size(); i++)
    {
        if(answer < vec[vec.size()-1][i])
            answer = vec[vec.size()-1][i];
    }
    
    
    return answer;
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.*;

class Solution {
    public int solution(int[][] triangle) {
        int answer = 0;
        
        int[][] dp = new int[triangle.length][triangle.length];
        dp[0][0] = triangle[0][0];
        
        
        for(int i=0; i<triangle.length-1; i++)
        {
            for(int t=0; t<triangle[i].length; t++)
            {
                dp[i+1][t] = Math.max(dp[i+1][t],dp[i][t]+triangle[i+1][t]);
                dp[i+1][t+1] = dp[i][t]+triangle[i+1][t+1];
            }
        }
        

        for(int t=0; t<dp[dp.length-1].length; t++)
        {
            answer = Math.max(answer, dp[dp.length-1][t]);
        }
        
        
        return answer;
    }
}